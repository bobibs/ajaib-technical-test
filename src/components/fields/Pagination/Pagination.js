import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import styles from './styles.module.css';

export default function Pagination(props) {
  const { className, onClick, resetValue } = props;
  console.log(props);
  const [currentPage, setCurrentPage] = useState(1);
  const [lastPage, setLastPage] = useState(5);

  const classes = [styles.root, className].filter(Boolean).join(' ');
  const pages = 50 / 10;
  const handleClick = (page) => {
    setCurrentPage(page);
    onClick(page);
  };
  const handleClickFirst = () => {
    setCurrentPage(1);
    onClick(1);
  };
  const handleClickPrev = () => {
    if (currentPage !== 1) {
      setCurrentPage(currentPage - 1);
      onClick(currentPage - 1);
    }
  };
  const handleClickNext = () => {
    if (currentPage !== 5) {
      setCurrentPage(currentPage + 1);
      onClick(currentPage + 1);
    }
  };
  const handleClickLast = () => {
    setCurrentPage(5);
    onClick(5);
  };

  useEffect(() => {
    if (resetValue) {
      setCurrentPage(1);
    }
  }, [resetValue]);

  return (
    <div className={classes}>
      <div className={styles.child} onClick={handleClickFirst}>
        First
      </div>
      <div className={styles.child} onClick={handleClickPrev}>
        Prev
      </div>
      <div className={styles.childs}>
        {Array.from(Array(pages).keys()).map((i, idx) => (
          <div
            className={`${styles.child} ${
              currentPage === idx + 1 ? styles.childActive : ''
            }`}
            key={idx}
            onClick={() => handleClick(i + 1)}
          >
            {i + 1}
          </div>
        ))}
      </div>
      <div className={styles.child} onClick={handleClickNext}>
        Next
      </div>
      <div className={styles.child} onClick={handleClickLast}>
        Last
      </div>
    </div>
  );
}

Pagination.defaultProps = {
  className: '',
  onClick: () => {},
  resetValue: false,
};

Pagination.propTypes = {
  className: PropTypes.string,
  onClick: PropTypes.func,
  resetValue: PropTypes.bool,
};
