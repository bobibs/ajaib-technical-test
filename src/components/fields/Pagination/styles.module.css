.root {
    display: flex;
}

.root > div:first-child {
    border-radius: 0.25rem 0 0 0.2rem;
    border-left: 0.0125rem solid rgba(17, 17, 17, 0.25);
}

.root > div:last-child {
    border-radius: 0 0.25rem 0.25rem 0;
}

.childs {
    display: flex;
}

.child {
    border-top: 0.0125rem solid rgba(17, 17, 17, 0.25);
    border-right: 0.0125rem solid rgba(17, 17, 17, 0.25);
    border-bottom: 0.0125rem solid rgba(17, 17, 17, 0.25);
    padding: 0.313rem 0.75rem;
    cursor: pointer;
    color: var(--primary-color);
}

.childActive,
.child:hover {
    background: var(--primary-color);
    color: #ffffff;
}
