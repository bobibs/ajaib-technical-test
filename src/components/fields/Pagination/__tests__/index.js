import React, { useEffect, useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Pagination from '../Pagination';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useState: jest.fn(),
  useEffect: jest.fn(),
}));

const currentPage = 1;
const setCurrentPage = jest.fn();
const lastPage = 5;
const setLastPage = jest.fn();

describe('src/components/elements/Pagination', () => {
  beforeEach(() => {
    useState.mockImplementation(() => [currentPage, setCurrentPage]);
    useState.mockImplementation(() => [lastPage, setLastPage]);
    useEffect.mockImplementation((fn) => fn());
  });

  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<Pagination resetValue={true} />);
    Pagination.defaultProps.onClick();
    expect(tree).toMatchSnapshot();
  });

  test('function handleClick', () => {
    const onClick = jest.fn();
    const result = Pagination({ onClick });

    result.props.children[0].props.onClick();
    expect(onClick).toHaveBeenCalledWith(1);

    result.props.children[1].props.onClick();
    expect(onClick).toHaveBeenCalledWith(1);

    result.props.children[2].props.children[0].props.onClick();
    expect(onClick).toHaveBeenCalledWith(1);

    result.props.children[3].props.onClick();
    expect(onClick).toHaveBeenCalledWith(1);

    result.props.children[4].props.onClick();
    expect(onClick).toHaveBeenCalledWith(1);
  });
});
