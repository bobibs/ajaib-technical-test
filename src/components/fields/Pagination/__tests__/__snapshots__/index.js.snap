// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/Pagination render 1`] = `
<div
  className="root"
>
  <div
    className="child"
    onClick={[Function]}
  >
    First
  </div>
  <div
    className="child"
    onClick={[Function]}
  >
    Prev
  </div>
  <div
    className="childs"
  >
    <div
      className="child "
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="child "
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="child "
      onClick={[Function]}
    >
      3
    </div>
    <div
      className="child "
      onClick={[Function]}
    >
      4
    </div>
    <div
      className="child childActive"
      onClick={[Function]}
    >
      5
    </div>
  </div>
  <div
    className="child"
    onClick={[Function]}
  >
    Next
  </div>
  <div
    className="child"
    onClick={[Function]}
  >
    Last
  </div>
</div>
`;
